### PR TITLE
WIP on FindReferences on property accessors

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/17684

### CODE REVIEW

**Comment by jcouv** [comment]
@CyrusNajmabadi I'm taking a look at https://github.com/dotnet/roslyn/issues/17684, but I'm having some trouble.

In short, I extended the property accessor finder to not just look for accessor methods by name ("get_Property") but also appropriate (read vs. write) usages of "Property". For example, `var x = Property` references the getter. 
Then I tried to avoid some duplication, by having the property finder avoid references already handled by the property accessor finder.

FindReferences on accessors works fairly well, but FindReferences on the property now reports separately on the property and its accessors.

Do you have any thoughts on how to handle this more cleanly?

```C#
interface I
{
    int {|Definition:Property|} { get; }
}
class Program : I
{
    public int {|Definition:Pro$$perty|} { {|Definition:get|}; set; }

    void M()
    {
        var i = [|Property|];
    }
}
```

image

image

For comparison, here's what this scenario looks like before my change:
image

**Comment by CyrusNajmabadi** [comment]
@jcouv This would *not* be my preference for how we expose this feature. Instead, i would like find-refs on a property to find all usages of it, but categorize them appropriately as reads, writes or reads-and-writes (i.e. Prop++).  

Note ReferenceLocation already has an IsWrittenTo property.  But we'd likely want to expand that.

Then, when presenting hte data, we'd have a column called something like "Accessed By"

Users could then see the reads/writes, and could even group by them if they wanted.  

In other words.  I don't think we should filter results.  Instead, i think this should be done at hte presentation level.

**Comment by CyrusNajmabadi** [comment]
As to your thoughts:

1. If we were to go down this route, i would remove the concept of finding properties in FindRefs entirely.  in other words, if you tried to FindRefs on a property, we we would preemptively make that a FindRefs on its getter and setter.  And we would never even report results for the property itself.

When searching, we would only use one finder, but on hte locations where it had a match, it woudl report finding a reference to the getter, setter, or both.

**Comment by CyrusNajmabadi** [comment]
So, in your example, i woudl only expect to see buckets for:

I.Property.get  and Program.Property.get   There would never be a bucket for I.Property

File: src/EditorFeatures/Test2/FindReferences/FindReferencesTests.vb [hunk 1]
```diff
@@ -222,12 +229,75 @@ Namespace Microsoft.CodeAnalysis.Editor.UnitTests.FindReferences
...
+            Dim builder = PooledStringBuilder.GetInstance()
+            builder.Builder.AppendLine()
+            If Not messageOnly Then
+                builder.Builder.AppendLine($"Expected: {String.Join(", ", expected.Select(Function(e) e.ToString()))}")
+                builder.Builder.AppendLine($"Actual: {String.Join(", ", actual.Select(Function(a) a.ToString()))}")
+            End If
+
+            Dim tree = doc.GetSyntaxTreeSynchronously(Nothing)
+            Dim position = 0
+            Dim sourceText As SourceText = tree.GetText()
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Test2/FindReferences/FindReferencesTests.vb]: can just get the text directly from the doc.  no need for the tree.

File: src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxFactsService.cs [hunk 2]
```diff
@@ -1304,6 +1304,54 @@ public bool IsLeftSideOfAnyAssignment(SyntaxNode node)
             return (node as ExpressionSyntax).IsLeftSideOfAnyAssignExpression();
         }
 
+        public bool IsAssignedTo(SyntaxNode node)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxFactsService.cs]: hrmm... we already have IsWrittenTo as an extension.  Do we need this?

Note: IsWrittenTo also handles hte cases where a VB property is written to if passed to a byref/out arg..

File: src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxFactsService.cs [hunk 3]
```diff
@@ -1304,6 +1304,54 @@ public bool IsLeftSideOfAnyAssignment(SyntaxNode node)
...
+            if (IsLeftSideOfAssignment(node))
+            {
+                // `node = ...`
+                return false;
+            }
+
+            if (node.Parent is MemberAccessExpressionSyntax access && access.Name == node)
+            {
+                // `x.node = ...`
+                return !IsLeftSideOfAssignment(access);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxFactsService.cs]: Note: you can have ```(x.node) = ...```

**Comment by jcouv** [comment]
Thanks for the feedback. I'll close the PR.
For the approach that you recommend (categorizing references during presentation), would you have a previous issue or PR that I could look like to get some idea?

**Comment by CyrusNajmabadi** [comment]
@jcouv Def don't close this yet :)   Get feedback from teh rest of the team.  :)

**Comment by CyrusNajmabadi** [comment]
> For the approach that you recommend (categorizing references during presentation)

I don't have a PR.  But what you can look at is:  https://github.com/dotnet/roslyn/blob/master/src/VisualStudio/Core/Def/Implementation/FindReferences/Entries/AbstractDocumentSpanEntry.cs#L52

An 'Entry' corresponds to a row in the FAR window.  This is basically the point where the UI calls into us to get data for the columns for that row.  

Now, in order to support this you'll have to:

1. change our internal data model to include the "Access type" enum value.  Our data model is https://github.com/dotnet/roslyn/blob/master/src/Features/Core/Portable/FindUsages/SourceReferenceItem.cs .  Note: TypeScript/F# use this as well.  So you need to make changes here in a compatible manner.  
2. Notify VS that you want an extra column in the list.  That will be done here i think:  https://github.com/dotnet/roslyn/blob/master/src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs#L100

Note: i'm not totally familiar on how you add columns.  If it's not immediately obvious/easy, talk to DaofaLi and he can point you in the right direction.

**Comment by CyrusNajmabadi** [comment]
Let me know if you need more information @jcouv .  i'd be happy to help as much as  i can on this.

**Comment by CyrusNajmabadi** [comment]
https://github.com/dotnet/roslyn/blob/master/src/Features/Core/Portable/FindUsages/SourceReferenceItem.cs#L11-L26

I imagine us adding this:

```c#
enum ReferenceType {
   None;
   Read = 1;
   Write = 2;
   // ... Future values ...
}
```

and

```c#
public ReferenceType ReferenceType { get; }
```

Ideally, we'd deprecate IsWrittenTo.  As this is an internal API, we can def do, and then wait for F#/TS/JS to move over to this new API.

**Comment by CyrusNajmabadi** [comment]
https://github.com/dotnet/roslyn/blob/715aa4a630246eceda0adf4683ed22cb7be70890/src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/AbstractReferenceFinder.cs#L231-L235 

this is where we'd then update all code to specify the reference type.
